docs/adles.1
.\" Man page generated from reStructuredText.
.
.TH "ADLES" "1" "Dec 13, 2017" "1.4.0" "ADLES"
.SH NAME
adles \- automated creation of virtualized educational environments
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
adles [options] [\-]
adles [options] [\-t TYPE] \-c SPEC [\-]
adles [options] (\-m | \-d) [\-p] \-s SPEC [\-]
adles [options] (–cleanup\-masters | –cleanup\-enviro) [–nets] \-s SPEC [\-]
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
Automated Deployment of Lab Environments System (ADLES)
.nf
ADLES automates the deterministic creation of virtualized environments for use in
Cybersecurity and Information Technology (IT) education.
The system enables educators to easily build deterministic and
portable environments for their courses, saving significant amounts of
time and effort, and alieviates the requirement of possessing advanced IT knowledge.
.fi
.sp
.sp
Complete documentation can be found at ReadTheDocs: \fI\%https://adles.readthedocs.io\fP
.sp
Project source code and examples can be found on GitHub: \fI\%https://github.com/GhostofGoes/ADLES\fP
.SH COMMAND LINE OPTIONS
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.TP
.B \-n\fP,\fB  \-\-no\-color
Do not color terminal output
.TP
.B \-v\fP,\fB  \-\-verbose
Emit debugging logs to terminal
.TP
.BI \-c\fP,\fB  \-\-validate \ SPEC
Validates syntax of an exercise specification
.TP
.BI \-t\fP,\fB  \-\-type \ TYPE
Type of specification to validate: exercise, package, infra
.TP
.BI \-s\fP,\fB  \-\-spec \ SPEC
Name of a YAML specification file
.TP
.BI \-i\fP,\fB  \-\-infra \ FILE
Override infra spec with the one in FILE
.TP
.B \-p\fP,\fB  \-\-package
Build environment from package specification
.TP
.B \-m\fP,\fB  \-\-masters
Master creation phase of specification
.TP
.B \-d\fP,\fB  \-\-deploy
Environment deployment phase of specification
.TP
.B \-\-cleanup\-masters
Cleanup masters created by a specification
.TP
.B \-\-cleanup\-enviro
Cleanup environment created by a specification
.TP
.B \-\-nets
Cleanup networks created during either phase
.TP
.BI \-\-print\-spec \ NAME
Prints the named specification: exercise, package, infrastructure
.TP
.B \-\-list\-examples
Prints the list of examples available
.TP
.BI \-\-print\-example \ NAME
Prints the named example
.TP
.B \-h\fP,\fB  \-\-help
Shows this help
.TP
.B \-\-version
Prints current version
.UNINDENT
.UNINDENT
.UNINDENT
.SH USAGE
.sp
Creating an environment using ADLES:
.INDENT 0.0
.IP \(bu 2
Read the exercise and infrastructure specifications and examples of them.
.IP \(bu 2
Write an infrastructure specification for your platform. (Currently, VMware vSphere is the only platform supported)
.IP \(bu 2
Write an exercise specification with the environment you want created.
.IP \(bu 2
Check its syntax, run the mastering phase, make your changes, and then run the deployment phase.
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
adles \-c my\-competition.yaml
adles \-m \-s my\-competition.yaml
adles \-d \-s my\-competition.yaml
.ft P
.fi
.UNINDENT
.UNINDENT
.SH EXAMPLES
.INDENT 0.0
.INDENT 3.5
adles –list\-examples
adles \-c examples/tutorial.yaml
adles –verbose –masters –spec examples/experiment.yaml
adles \-vds examples/competition.yaml
adles –cleanup\-masters –nets \-s examples/competition.yaml
adles –print\-example competition | adles \-v \-c \-
.UNINDENT
.UNINDENT
.SH LICENSING
.sp
This project is licensed under the Apache License, Version 2.0. See
LICENSE for the full license text, and NOTICES for attributions to
external projects that this project uses code from.
.SH PROJECT HISTORY
.sp
The system began as a proof of concept implementation of my Master’s thesis research at the
University of Idaho in Fall of 2016. It was originally designed to run on the RADICL lab.
.SH AUTHOR
Christopher Goes
.SH COPYRIGHT
2016 - 2017, Christopher Goes
.\" Generated by docutils manpage writer.
.


adles/scripts/adles.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# http://multivax.com/last_question.html

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""ADLES: Automated Deployment of Lab Environments System.
Uses YAML specifications to manage and create virtual environments.

Usage:
    adles.py [options] --check-syntax FILE
    adles.py [options] (--create-masters | --deploy) -s FILE
    adles.py [options] --package-dir NAME
    adles.py [options] (--cleanup-masters | --cleanup-enviro) [--network-cleanup] -s FILE
    adles.py --version
    adles.py (-h | --help)

Options:
    -h, --help                  Shows this help
    --version                   Prints current version
    --no-color                  Do not color terminal output
    -v, --verbose               Emit debugging logs to terminal
    -c, --check-syntax FILE     Validates syntax is valid per specification
    -s, --spec FILE             YAML file with the environment specification
    -p, --package-dir NAME      Filepath of the exercise package directory
    -m, --create-masters        Master creation phase of specification
    -d, --deploy                Environment deployment phase of specification
    --cleanup-masters           Cleanup masters created by a specification
    --cleanup-enviro            Cleanup environment created by a specification
    --network-cleanup           Cleanup networks created during either phase

Examples:
    adles.py --check-syntax examples/tutorial.yaml
    adles.py -v --create-masters --spec examples/experiment.yaml
    adles.py -d -s examples/competition.yaml
    adles.py --cleanup-masters --network-cleanup -s examples/competition.yaml

"""

import logging

from docopt import docopt

from adles.interfaces import Interface
from adles.parser import check_syntax
from adles.utils import setup_logging
from adles import __version__


args = docopt(__doc__, version=__version__, help=True)

colors = (False if args["--no-color"] else True)
setup_logging(filename='adles.log', colors=colors, console_verbose=args["--verbose"])

if args["--spec"]:
    spec = check_syntax(args["--spec"])
    if spec is None:
        logging.error("Syntax check failed")
        exit(0)

    interface = Interface(spec)
    if args["--create-masters"]:
        interface.create_masters()
        logging.info("Finished creation of Masters for environment %s",
                     spec["metadata"]["name"])
    elif args["--deploy"]:
        interface.deploy_environment()
        logging.info("Finished deployment of environment %s", spec["metadata"]["name"])
    elif args["--cleanup-masters"]:
        interface.cleanup_masters(args["--network-cleanup"])
    elif args["--cleanup-enviro"]:
        interface.cleanup_environment(args["--network-cleanup"])
    else:
        logging.error("Invalid flags for --spec. Argument dump:\n%s", str(args))

elif args["--check-syntax"]:
    if check_syntax(args["--check-syntax"]) is None:
        logging.error("Syntax check failed")

elif args["--package-dir"]:
    logging.error("PACKAGES ARE CURRENTLY UNSUPPORTED")

else:
    logging.error("Invalid arguments. Argument dump:\n%s", str(args))

exit(0)
